File: 735.asteroid-collision.py
#
# @lc app=leetcode id=735 lang=python3
#
# [735] Asteroid Collision
#
from typing import List

# @lc code=start
class Solution:
    # T: O(n)
    # S: O(1) excl. res
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        res = []
        for x in asteroids:
            if x > 0:
                # e.g. [a, b], a < 0, b > 0, never collide
                res.append(x)
            else:
                while len(res) > 0 and res[-1] > 0 and res[-1] < -x:
                    # e.g. [a, b], a > 0, b < 0, abs(a) < abs(b)  
                    res.pop()
                
                if len(res) == 0 or res[-1] < 0:
                    # Two asteroids moving in the same direction will never meet
                    res.append(x)
                elif res[-1] == -x:
                    # Same size, opposite directions, both explode
                    res.pop()
        
        return res
        
# @lc code=end

File: 735.asteroid-collision.cpp
/*
 * @lc app=leetcode id=735 lang=cpp
 *
 * [735] Asteroid Collision
 */
#include <bits/stdc++.h>

// @lc code=start
class Solution {
public:
    // Stack simulation
    // T: O(n)
    // S: O(n) (may be considered O(1) because the result container and the stack are the same)
    std::vector<int> asteroidCollision(std::vector<int>& asteroids) {
        std::vector<int> res;  // also simulates the stack
        res.reserve(asteroids.size());
        for (const int& a : asteroids) {
            if (a > 0) {
                res.push_back(a);
            } else {    // a < 0
                // Colliding (except for equal size case)
                while (!res.empty() && res.back() > 0 && res.back() < -a) {
                    res.pop_back();
                }
                if (res.empty() || res.back() < 0) {
                    // all asteroids traveling left
                    res.push_back(a);
                } else if (res.back() == -a) {
                    // If abs values equal, can only pop one, so this case 
                    // is separate from the while loop above
                    res.pop_back();
                }
            }           
        }
        return res;
    }
};
// @lc code=end

